spanner_dbapi/connection.py

```diff
@@ -5,66 +5,145 @@
+                       transactions must be committed.
+
+    :type read_only: bool
+    :param read_only: (Optional) Indicates if the Connection is intended to be
+                      used for read-only transactions.
+    """
+
+    def __init__(self, database, instance, autocommit=True, read_only=False):
+        self.database = database
+        self.instance = instance
+        self.autocommit = autocommit
+        self.read_only = read_only
+        self.transaction_mode = (
+            TransactionModes.read_only
+            if self.read_only
+            else TransactionModes.read_write
+        )
+        self.autocommit_dml_mode = AutocommitDMLModes.transactional
+        self.timeout_secs = 0
+        self.read_timestamp = None
+        self.commit_timestamp = None
+        self.__is_closed = False
+        self.__inside_transaction = not autocommit
+        self.__transaction_started = False
+        self.__ddl_statements = []
+        self.read_only_staleness = {}
+
+    @property
+    def is_closed(self):
+        return self.__is_closed
+
+    @property
+    def inside_transaction(self):
+        return self.__inside_transaction
+
+    @property
+    def transaction_started(self):
+        return self.__transaction_started
+
+    @property
+    def _ddl_statements(self):
+        return self.__ddl_statements
+
+    @_ddl_statements.setter
+    def _ddl_statements(self, dll):
+        self._change_transaction_started(True)
+
+        self.__ddl_statements = dll
+        if self.autocommit:
+            self.commit()
+
+    def _change_transaction_started(self, val: bool):
+        if self.__inside_transaction:
+            self.__transaction_started = val
+
+    @_is_conn_closed_check
     def __handle_update_ddl(self, ddl_statements):
+        """Run the list of Data Definition Language (DDL) statements on the
+        underlying database. Each DDL statement MUST NOT contain a semicolon.
+
+        :type ddl_statements: list
+        :param ddl_statements: A list of DDL statements, each without a
+                               semicolon.
+
+        :rtype: :class:`google.api_core.operation.Operation`
+        :returns: an operation instance
         """
-        Run the list of Data Definition Language (DDL) statements on the underlying
-        database. Each DDL statement MUST NOT contain a semicolon.
-        Args:
-            ddl_statements: a list of DDL statements, each without a semicolon.
-        Returns:
-            google.api_core.operation.Operation.result()
-        """
-        self.__raise_if_already_closed()
         # Synchronously wait on the operation's completion.
-        return self._dbhandle.update_ddl(ddl_statements).result()
+        return self.database.update_ddl(ddl_statements).result()
 
+    @_is_conn_closed_check
     def read_snapshot(self):
-        self.__raise_if_already_closed()
-        return self._dbhandle.snapshot()
+        return self.database.snapshot()
```

Comment:
What is the specification that requires this method? If it's not listed in the design document or any other relevant specification sheet, or discussed and approved a priori by the Spanner team, then there is no reason to have it here at all.